File: backend/src/models/TinTuyenDung.js
const mongoose = require('mongoose');
const { TRANG_THAI_TIN } = require('../utils/constants');

const tinTuyenDungSchema = new mongoose.Schema({
  tieuDe: {
    type: String,
    required: [true, 'Vui lòng nhập tiêu đề tin tuyển dụng'],
    trim: true,
    maxlength: [300, 'Tiêu đề không được quá 300 ký tự']
  },
  slug: {
    type: String,
    unique: true,
    lowercase: true
  },
  moTa: {
    type: String,
    required: [true, 'Vui lòng nhập mô tả công việc'],
    trim: true
  },
  yeuCau: {
    type: String,
    required: [true, 'Vui lòng nhập yêu cầu ứng viên'],
    trim: true
  },
  quyenLoi: {
    type: String,
    trim: true
  },
  kyNangBatBuoc: [{
    type: String,
    trim: true
  }],
  kyNangMongMuon: [{
    type: String,
    trim: true
  }],
  kinhNghiemToiThieu: {
    type: Number,
    default: 0,
    min: [0, 'Kinh nghiệm không được âm']
  },
  capBac: {
    type: String,
    enum: ['Thuc_tap', 'Nhan_vien', 'Truong_nhom', 'Quan_ly', 'Giam_doc'],
    default: 'Nhan_vien'
  },
  hinhThucLamViec: {
    type: String,
    enum: ['Toan_thoi_gian', 'Ban_thoi_gian', 'Thuc_tap', 'Hop_dong'],
    default: 'Toan_thoi_gian'
  },
  mucLuong: {
    min: {
      type: Number,
      default: 0
    },
    max: {
      type: Number,
      default: 0
    },
    donVi: {
      type: String,
      default: 'VND'
    },
    hienThi: {
      type: Boolean,
      default: true
    }
  },
  diaDiem: {
    type: String,
    required: [true, 'Vui lòng nhập địa điểm làm việc'],
    trim: true
  },
  soLuongTuyen: {
    type: Number,
    required: [true, 'Vui lòng nhập số lượng tuyển'],
    min: [1, 'Số lượng tuyển phải lớn hơn 0']
  },
  soLuongDaTuyen: {
    type: Number,
    default: 0,
    min: 0
  },
  hanNop: {
    type: Date,
    required: [true, 'Vui lòng nhập hạn nộp hồ sơ']
  },
  quyTrinhId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'QuyTrinhTuyenDung',
    required: true
  },
  trangThai: {
    type: String,
    enum: Object.values(TRANG_THAI_TIN),
    default: TRANG_THAI_TIN.NHAP
  },
  taoBoi: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung',
    required: true
  },
  ngayDang: {
    type: Date,
    default: null
  },
  luotXem: {
    type: Number,
    default: 0
  },
  soLuongUngTuyen: {
    type: Number,
    default: 0
  }
}, {
  timestamps: true,
  collection: 'tin_tuyen_dung'
});

// Index
tinTuyenDungSchema.index({ slug: 1 });
tinTuyenDungSchema.index({ trangThai: 1 });
tinTuyenDungSchema.index({ hanNop: 1 });
tinTuyenDungSchema.index({ taoBoi: 1 });
tinTuyenDungSchema.index({ createdAt: -1 });

// Virtual để kiểm tra còn hạn
tinTuyenDungSchema.virtual('conHan').get(function() {
  return new Date() < new Date(this.hanNop);
});

// Virtual để kiểm tra đã tuyển đủ
tinTuyenDungSchema.virtual('daTuyenDu').get(function() {
  return this.soLuongDaTuyen >= this.soLuongTuyen;
});

// Tự động tạo slug từ tiêu đề
tinTuyenDungSchema.pre('save', async function(next) {
  if (this.isModified('tieuDe') && !this.slug) {
    const { createSlug } = require('../utils/helpers');
    let slug = createSlug(this.tieuDe);
    
    // Kiểm tra trùng lặp
    let slugExists = await this.constructor.findOne({ slug });
    let counter = 1;
    
    while (slugExists) {
      slug = `${createSlug(this.tieuDe)}-${counter}`;
      slugExists = await this.constructor.findOne({ slug });
      counter++;
    }
    
    this.slug = slug;
  }
  
  // Set ngày đăng khi chuyển sang trạng thái Dang_tuyen
  if (this.isModified('trangThai') && this.trangThai === TRANG_THAI_TIN.DANG_TUYEN && !this.ngayDang) {
    this.ngayDang = new Date();
  }
  
  next();
});

const TinTuyenDung = mongoose.model('TinTuyenDung', tinTuyenDungSchema);

module.exports = TinTuyenDung;

File: frontend/lib/models/tin_tuyen_dung.dart
class MucLuong {
  final int min;
  final int max;
  final String donVi;
  final bool hienThi;

  MucLuong({
    required this.min,
    required this.max,
    this.donVi = 'VND',
    this.hienThi = true,
  });

  factory MucLuong.fromJson(Map<String, dynamic> json) {
    return MucLuong(
      min: json['min'] ?? 0,
      max: json['max'] ?? 0,
      donVi: json['donVi'] ?? 'VND',
      hienThi: json['hienThi'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'min': min,
      'max': max,
      'donVi': donVi,
      'hienThi': hienThi,
    };
  }
}

class TinTuyenDung {
  final String id;
  final String tieuDe;
  final String slug;
  final String moTa;
  final String yeuCau;
  final String? quyenLoi;
  final List<String> kyNangBatBuoc;
  final List<String> kyNangMongMuon;
  final int kinhNghiemToiThieu;
  final String capBac;
  final String hinhThucLamViec;
  final MucLuong mucLuong;
  final String diaDiem;
  final int soLuongTuyen;
  final int soLuongDaTuyen;
  final DateTime hanNop;
  final String quyTrinhId;
  final String trangThai;
  final String taoBoi;
  final DateTime? ngayDang;
  final int luotXem;
  final int soLuongUngTuyen;
  final DateTime createdAt;
  final DateTime updatedAt;

  TinTuyenDung({
    required this.id,
    required this.tieuDe,
    required this.slug,
    required this.moTa,
    required this.yeuCau,
    this.quyenLoi,
    required this.kyNangBatBuoc,
    required this.kyNangMongMuon,
    required this.kinhNghiemToiThieu,
    required this.capBac,
    required this.hinhThucLamViec,
    required this.mucLuong,
    required this.diaDiem,
    required this.soLuongTuyen,
    required this.soLuongDaTuyen,
    required this.hanNop,
    required this.quyTrinhId,
    required this.trangThai,
    required this.taoBoi,
    this.ngayDang,
    required this.luotXem,
    required this.soLuongUngTuyen,
    required this.createdAt,
    required this.updatedAt,
  });

  factory TinTuyenDung.fromJson(Map<String, dynamic> json) {
    return TinTuyenDung(
      id: json['_id'] ?? '',
      tieuDe: json['tieuDe'] ?? '',
      slug: json['slug'] ?? '',
      moTa: json['moTa'] ?? '',
      yeuCau: json['yeuCau'] ?? '',
      quyenLoi: json['quyenLoi'],
      kyNangBatBuoc: List<String>.from(json['kyNangBatBuoc'] ?? []),
      kyNangMongMuon: List<String>.from(json['kyNangMongMuon'] ?? []),
      kinhNghiemToiThieu: json['kinhNghiemToiThieu'] ?? 0,
      capBac: json['capBac'] ?? 'Nhan_vien',
      hinhThucLamViec: json['hinhThucLamViec'] ?? 'Toan_thoi_gian',
      mucLuong: MucLuong.fromJson(json['mucLuong'] ?? {}),
      diaDiem: json['diaDiem'] ?? '',
      soLuongTuyen: json['soLuongTuyen'] ?? 1,
      soLuongDaTuyen: json['soLuongDaTuyen'] ?? 0,
      hanNop: DateTime.parse(json['hanNop']),
      quyTrinhId: json['quyTrinhId'] ?? '',
      trangThai: json['trangThai'] ?? 'Nhap',
      taoBoi: json['taoBoi'] ?? '',
      ngayDang: json['ngayDang'] != null ? DateTime.parse(json['ngayDang']) : null,
      luotXem: json['luotXem'] ?? 0,
      soLuongUngTuyen: json['soLuongUngTuyen'] ?? 0,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'tieuDe': tieuDe,
      'slug': slug,
      'moTa': moTa,
      'yeuCau': yeuCau,
      'quyenLoi': quyenLoi,
      'kyNangBatBuoc': kyNangBatBuoc,
      'kyNangMongMuon': kyNangMongMuon,
      'kinhNghiemToiThieu': kinhNghiemToiThieu,
      'capBac': capBac,
      'hinhThucLamViec': hinhThucLamViec,
      'mucLuong': mucLuong.toJson(),
      'diaDiem': diaDiem,
      'soLuongTuyen': soLuongTuyen,
      'soLuongDaTuyen': soLuongDaTuyen,
      'hanNop': hanNop.toIso8601String(),
      'quyTrinhId': quyTrinhId,
      'trangThai': trangThai,
      'taoBoi': taoBoi,
      'ngayDang': ngayDang?.toIso8601String(),
      'luotXem': luotXem,
      'soLuongUngTuyen': soLuongUngTuyen,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  // Helper methods
  bool get conHan => DateTime.now().isBefore(hanNop);
  bool get daTuyenDu => soLuongDaTuyen >= soLuongTuyen;
  bool get dangTuyen => trangThai == 'Dang_tuyen';
  
  String get mucLuongText {
    if (!mucLuong.hienThi) return 'Thỏa thuận';
    if (mucLuong.min == 0 && mucLuong.max == 0) return 'Thỏa thuận';
    
    String formatMoney(int amount) {
      if (amount >= 1000000) {
        return '${(amount / 1000000).toStringAsFixed(0)} triệu';
      }
      return '${amount ~/ 1000}K';
    }
    
    if (mucLuong.min == mucLuong.max) {
      return formatMoney(mucLuong.min);
    }
    return '${formatMoney(mucLuong.min)} - ${formatMoney(mucLuong.max)}';
  }
}